src/mrinufft/trajectories/tools.py

```diff
@@ -369,6 +380,290 @@ def unepify(trajectory: NDArray, Ns_readouts: int, Ns_transitions: int) -> NDArr
+    smax: float = DEFAULT_SMAX,
+    n_jobs: int = 1,
+) -> tuple[NDArray, NDArray, NDArray, NDArray]:
+    """Get gradient timing values for trapezoidal or triangular waveforms.
+
+    Compute gradient timing values to take k-space trajectories
+    from position ``ks`` with gradient ``gs`` to position ``ke`` with gradient
+    ``ge``, while being hardware compliant.
+    This function calculates the minimal number of time steps required for
+    the ramp down, ramp up, and plateau phases of the gradient waveform,
+    ensuring that the area traversed in k-space matches the desired
+    trajectory while adhering to the maximum gradient amplitude and
+    slew rate constraints.
+
+    Parameters
+    ----------
+    kspace_end_loc : NDArray
+        Ending k-space positions, shape (nb_shots, nb_dimension).
+    kspace_start_loc : NDArray, default None when it is 0
+        Starting k-space positions, shape (nb_shots, nb_dimension).
+    end_gradients : NDArray, default None when it is 0
+        Ending gradient values, shape (nb_shots, nb_dimension).
+    start_gradients : NDArray, default None when it is 0
+        Starting gradient values, shape (nb_shots, nb_dimension).
+    gamma : float, optional
+        Gyromagnetic ratio in Hz/T. Default is Gammas.Hydrogen.
+    raster_time : float, optional
+        Time interval between gradient samples (s). Default is DEFAULT_RASTER_TIME.
+    gmax : float, optional
+        Maximum gradient amplitude (T/m). Default is DEFAULT_GMAX.
+    smax : float, optional
+        Maximum slew rate ``T/m/s``. Default is DEFAULT_SMAX.
+    n_jobs : int, optional
+        Number of parallel jobs to run for optimization, by default 1.
+
+    Returns
+    -------
+    n_ramp_down: The timing values for the ramp down phase.
+    n_ramp_up: The timing values for the ramp up phase.
+    n_plateau: The timing values for the plateau phase.
+    gi: The intermediate gradient values for trapezoidal or triangular waveforms.
+    """
+    area_needed = (kspace_end_loc - kspace_start_loc) / gamma / raster_time
+
+    def solve_gi_min_plateau(gs, ge, area):
+        def _residual(gi):
+            n_down, n_up = _calculate_ramps(gs, ge, gi, smax, raster_time)
+            n_pl = _calculate_plateau(gs, ge, gi, n_down, n_up, area)
+            if n_pl < 0:
+                return np.abs(n_pl) * 10000  # Penalize negative plateau
+            return n_pl * 100
+
+        res = minimize_scalar(
+            _residual,
+            bounds=(-gmax, gmax),
+            method="bounded",
+        )
+        if not res.success:
+            raise RuntimeError(f"Minimization failed: {res.message}")
+        return res.x
+
+    gi = Parallel(n_jobs=n_jobs)(
+        delayed(solve_gi_min_plateau)(
+            start_gradients[i, j],
+            end_gradients[i, j],
+            area_needed[i, j],
+        )
+        for i in range(start_gradients.shape[0])
+        for j in range(start_gradients.shape[1])
+    )
```

Comment:
```suggestion
    gi = Parallel(n_jobs=n_jobs)(
        delayed(solve_gi_min_plateau)(gs,ge,area) 
        for gs,ge,area in zip(start_gradients[:],end_gradients[:], area_needed[:]))
```
more pythonic :) I am almost sure that a full vectorized method is also possible, but I don't deem this to be too critical. 

The main question is if this takes some non-negligeable times or not ? (e.g. is joblib really necessary here)